scipy/sparse/construct.py

```diff
@@ -618,8 +618,8 @@ def bmat(blocks, format=None, dtype=None):
         B = blocks[i, j]
         idx = slice(nnz, nnz + B.nnz)
         data[idx] = B.data
-        row[idx] = B.row + row_offsets[i]
-        col[idx] = B.col + col_offsets[j]
+        row[idx] = np.dtype(idx_dtype).type(B.row) + row_offsets[i]
+        col[idx] = np.dtype(idx_dtype).type(B.col) + col_offsets[j]
```

Comment:
This casting is a bit strange looking to my eye. How about this:

```
np.add(B.row, row_offsets[i], out=row[idx], dtype=idx_dtype)
np.add(B.col, col_offsets[i], out=col[idx], dtype=idx_dtype)
```

This is a bit more explicit about what's happening regarding the dtypes, and may be slightly more efficient as well.

Reply:
Agreed that the original change here looks a little odd. @kevinrichardgreen could you make this change and resolve the merge conflict? (if you don't know how to do that, please let me know and I'll help)

Reply:
Changed. Though the merge really does seem to be a mess right now. I've established that this just need to be moved into _construct.py. I can force push my local branch (that is rebased off of scipy/master) that has the desired version to kevinrichardgreen:fix-bmat so that it can be accepted from this PR. (Sidenote: do forced pushes mess with PRs in any substantial way?)

Reply:
Thanks @kevinrichardgreen!

> Sidenote: do forced pushes mess with PRs in any substantial way?

No, that is perfectly fine. A PR always just reflects whatever the state is of the remote branch that the PR was originally made from. You can always rewrite history however you like.